README.rst
Introduction
============

.. image:: https://badge.fury.io/py/bottle-gui.png
    :target: http://badge.fury.io/py/bottle-gui

.. image:: https://pypip.in/d/bottle-gui/badge.png
        :target: https://crate.io/packages/bottle-gui?version=latest

This package is used to vizualize services in `Bottle <http://bottlepy.org>`_
web framework.

Documentation
-------------

Full module documentation and description can be found at Read the Docs:

- http://bottle-gui.readthedocs.org/


CHANGES.rst
Changelog
=========

0.2.1
-----
    - Small package metadata improvements.

0.2.0
-----
    - Added support for blacklist (.BLACKLIST property). See tests for details.
    - ``bottle_gui()`` function renamed to just ``gui()``. There was package named `bottle_gui` with source `bottle_gui.py`` with function `bottle_gui`. This is imho better.

0.1.3
-----
    - Paths are now sorted alphabetically.

0.1.2
-----
    - Fixed bug in MANIFEST.in.

0.1.0
-----
    - Project created.


tests/test_gui.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
# This work is licensed under a Creative Commons 3.0 Unported License
# (http://creativecommons.org/licenses/by/3.0/).
#
# Imports ====================================================================
import sys
import time
import random
from multiprocessing import Process

import pytest
import requests
from bottle import run

import services  # local services for test purposes

sys.path.insert(0, 'src')
import bottle_gui


# Variables ==================================================================
ADDR = "127.0.0.1"
PORT = random.randint(10000, 65525)
URL = 'http://%s:%s%s' % (ADDR, PORT, "/")
SERV = None


# Functions & classes ========================================================
def run_server():
    bottle_gui.bottle_gui.BLACKLIST = []
    bg = bottle_gui.gui()
    run(
        host=ADDR,
        port=PORT,
        # server="paste",
        # debug=True,
        # reloader=True,
        # ssl_pem="static/host.pem"
    )


# Setup/Teardown ==============================================================
def setup_module(module):
    """
    Run server as subprocess.

    This is something like module constructor.
    """
    global SERV
    SERV = Process(target=run_server)
    SERV.start()
    time.sleep(1)


def teardown_module(module):
    """
    Shut down server subprocess.

    This is something like module destructor.
    """
    SERV.terminate()

    # test that server was really closed
    with pytest.raises(requests.exceptions.ConnectionError):
        requests.get(URL)


# Tests =======================================================================
def test_html_output():
    res = requests.get(URL)
    assert "<title>API index</title>" in res.text

    assert "Xex module docstring." in res.text
    assert "/sources/xex" in res.text
    assert "Another docstring" in res.text

    assert "Hist module level docstring." in res.text
    assert "/sources/hist" in res.text
    assert "/sources/hist/xe" in res.text
    assert "Here is hist docstring and so on." in res.text
    assert "Here is hist/xe docstring and so on." in res.text


def test_json_output():
    res = requests.get(URL, headers={'Accept': 'text/json'})
    data = res.json()

    root = {
        u'/': [
            {
                u'args': [],
                u'docstring': u'Handle requests to root of the project.',
                u'path': u'/',
                u'mdocstring': None,
                u'module_name': u'bottle_gui.bottle_gui',
                u'method': u'GET'
            }
        ]
    }

    static = {
        u'/bottle_gui_static/': [
            {
                u'args': [u'fn'],
                u'docstring': u'Serve static files.',
                u'path': u'/bottle_gui_static/',
                u'mdocstring': None,
                u'module_name': u'bottle_gui.bottle_gui',
                u'method': u'GET'
            }
        ]
    }

    hist = {
        u'/sources/hist': [
            {
                u'args': [u'something', u'something_else'],
                u'docstring': u'Here is hist/xe docstring and so on.',
                u'path': u'/sources/hist/xe',
                u'mdocstring': u'Hist module level docstring.',
                u'module_name': u'services.hist',
                u'method': u'GET'
            },
            {
                u'args': [u'something', u'something_else'],
                u'docstring': u'Here is hist docstring and so on.',
                u'path': u'/sources/hist',
                u'mdocstring': u'Hist module level docstring.',
                u'module_name': u'services.hist',
                u'method': u'GET'
            }
        ]
    }

    xex = {
        u'/sources/xex': [
            {
                u'args': [u'something', u'something_else'],
                u'docstring': u'Another docstring',
                u'path': u'/sources/xex',
                u'mdocstring': u'Xex module docstring.',
                u'module_name': u'services.xex',
                u'method': u'GET'
            }
        ]
    }

    assert root in data
    assert static in data
    assert hist in data
    assert xex in data


tests/services/xex.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================,
"""
Xex module docstring.
"""
from bottle import route


# Functions & classes =========================================================
@route("/sources/xex")
def test2(something, something_else):
    """
    Another docstring
    """
    pass


tests/services/hist.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
"""
Hist module level docstring.
"""
from bottle import route


# Functions & classes =========================================================
@route("/sources/hist")
def hist(something, something_else):
    """
    Here is hist docstring and so on.
    """
    pass


@route("/sources/hist/xe")
def hist_xe(something, something_else):
    """
    Here is hist/xe docstring and so on.
    """
    pass


src/bottle_gui/bottle_gui.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
import json
import inspect
import os.path
from string import Template

import bottle
from bottle import route, static_file, request, response

from napoleon2html import napoleon_to_html


# Variables ===================================================================
TEMPLATE_PATH = "static/templates/"  #: Path to the template directory.


def read_template(template_name):
    """
    Read content of the template file.

    Args:
        template_name (str): Name of the file.

    Returns:
        str: Content of `template_name` from :attr:`TEMPLATE_PATH` directory.
    """
    template_path = os.path.join(
        os.path.dirname(__file__),
        TEMPLATE_PATH,
        template_name
    )

    with open(template_path) as f:
        return f.read()


# load all necessary templates
INDEX_TEMPLATE = read_template("index.html")  #: static/templates/index.html
TABLE_TEMPLATE = read_template("table.html")  #: static/templates/table.html
ROW_TEMPLATE   = read_template("row.html")  #: static/templates/row.html
DESCR_TEMPLATE = read_template("descr.html")  #: static/templates/descr.html
BLACKLIST = ["/", "/bottle_gui_static/"]


# Classes =====================================================================
class RouteInfo(object):
    """
    Container for informations about `route`.

    Attributes:
        method (fn reference): Reference to undecorated function.
        path (str): Path to the function in bottle.
        args (list): Args of the function.
        docstring (str): Docstring for the function.
        mdocstring (str): Docstring for the module where the function is.
        module_name (str): Name of the module where the function is.
    """
    def __init__(self, method, path, args, docstring, mdocstring, module_name):
        """
        Attributes
            method (fn reference): see Attributes section for details.
            path (str): see Attributes section for details.
            args (list): see Attributes section for details.
            docstring (str): see Attributes section for details.
            mdocstring (str): see Attributes section for details.
            module_name (str): see Attributes section for details.
        """
        self.method = method
        self.path = path
        self.args = args
        self.docstring = self._sanitize(docstring)
        self.mdocstring = self._sanitize(mdocstring)
        self.module_name = module_name

    def _sanitize(self, s):
        """
        Replace ``<`` and ``>`` with corresponding HTML entities.

        Args:
            s (str): Input string.

        Returns:
            str: String with entities, or `s` ``if not s``.
        """
        if s:
            return s.replace("<", "&lt;").replace(">", "&gt;")

        return s

    def to_html(self):
        """
        Convert informations about this route to HTML.

        Note:
            :attr:`DESCR_TEMPLATE` and :attr:`ROW_TEMPLATE` is used.

        Returns:
            str: HTML representation of the `route`.
        """
        descr = ""

        # process docstring
        if self.docstring:
            docstring = self.docstring.strip() or ""

            descr = Template(DESCR_TEMPLATE).substitute(
                method_description=napoleon_to_html(docstring)
            )

        # wrap arguments to the html
        args = self.args or ""
        if args:
            args_style = "&#8672; &lt;<span class='param'>"
            args = args_style + "</span>, <span class='param'>".join(args)
            args += "</span>&gt;"

        return Template(ROW_TEMPLATE).substitute(
            name=self.path,
            args=args,
            http_type=self.method,
            method_description=descr
        )

    def to_dict(self):
        """
        Return dictionary representation of the class. This method is used for
        JSON output.

        Returns:
            dict: Dictionary following keys: ``method``, ``path``, ``args``, \
                  ``docstring``, ``mdocstring``, ``module_name``.
        """
        return {
            "method": self.method,
            "path": self.path,
            "args": self.args,
            "docstring": self.docstring,
            "mdocstring": self.mdocstring,
            "module_name": self.module_name,
        }

    def __str__(self):  # TODO: remove
        return self.method + " " + self.path


class RouteGroup(object):
    """
    This object is used to group :class:`RouteInfo` objects.

    Args:
        routes (list, default []): List with :class:`RouteInfo` objects.
    """
    def __init__(self, routes=[]):
        self.routes = routes

    def get_path(self):  # TODO: shortest path
        """
        Return `path` for this group.

        Returns:
            str: Path.
        """
        if len(self.routes) == 1:
            return self.routes[0].path

        # get longest path
        route_paths = map(lambda x: x.path, self.routes)
        longest_path = max(route_paths)

        if longest_path.endswith("/"):
            return longest_path

        return os.path.dirname(longest_path)

    def get_docstring(self):
        """
        Return 'module' docstring.

        Returns:
            str: Module docstring, if defined, or blank string.
        """
        if self.routes:
            return self.routes[0].mdocstring or ""

        return ""

    def to_html(self):
        """
        Convert group and all contained paths to HTML.

        Note:
            :attr:`TABLE_TEMPLATE` is used.

        Returns:
            str: HTML.
        """
        return Template(TABLE_TEMPLATE).substitute(
            name=self.get_path(),
            description=self.get_docstring(),
            rows="\n".join(
                map(
                    lambda x: x.to_html(),
                    sorted(self.routes, key=lambda x: x.path)
                )
            )
        )

    def to_dict(self):
        """
        Convert group to dict. This method is used for JSON output.

        Returns:
            dict: {path: [routes]}

        See Also:
            :meth:`RouteInfo.to_dict`
        """
        return {
            self.get_path(): map(lambda x: x.to_dict(), self.routes)
        }

    def __str__(self):  # TODO: remove
        return "group: " + " ".join(map(lambda x: str(x), self.routes)) + "\n"


# Functions ===================================================================
def list_routes():
    """
    Get list of :class:`RouteInfo` objects from bottle introspection.

    Returns:
        list: :class:`RouteInfo` objects.
    """
    return map(
        lambda r: RouteInfo(
            method=r.method,
            path=r.rule.split("<")[0],
            args=r.get_callback_args(),
            docstring=inspect.getdoc(r.get_undecorated_callback()) or "",
            mdocstring=inspect.getdoc(
                inspect.getmodule(r.get_undecorated_callback())
            ),
            module_name=r.get_undecorated_callback().__module__
        ),
        bottle.default_app().routes
    )


def group_routes(ungrouped_routes):
    """
    Group list of :class:`RouteInfo` objects in `ungrouped_routes` by their
    :attr:`RouteInfo.path` properties.

    Args:
        ungrouped_routes (list): List of :class:`RouteInfo` objects.

    Returns:
        list: :class:`RouteGroup` objects.
    """
    groups = []

    # go from longest routes to shorter
    routes = sorted(
        ungrouped_routes,
        key=lambda x: len(x.path),
        reverse=True
    )

    # group and remove / routes - they would break grouping algorithm later
    root_paths = filter(lambda x: x.path == "/", routes)
    if root_paths:
        groups.append(
            RouteGroup(root_paths)
        )

        # remove / routes if present
        map(lambda x: routes.pop(), root_paths)

    processed = set()
    for route in routes:
        # skip already processed routes
        if route in processed:
            continue

        # find all routes, which starts with `route`
        same_group = filter(
            lambda x: x.path.startswith(route.path),
            routes
        )

        # skip routes without groups for later processing
        if len(same_group) <= 1:  # contains always actual route, so <= 1
            continue

        groups.append(RouteGroup(same_group))
        processed.update(same_group)

    # don't forget to uprocessed routes
    for route in set(routes) - processed:
        groups.append(
            RouteGroup([route])
        )

    return groups


def to_html(grouped_routes):
    """
    Convert list of :class:`RouteGroup` objects in `group_routes` to HTML.

    Args:
        grouped_routes (list): Llist of :class:`RouteGroup` objects.

    Returns:
        str: HTML page with routes.
    """
    return Template(INDEX_TEMPLATE).substitute(
        tables="\n".join(
            map(
                lambda x: x.to_html(),
                sorted(grouped_routes, key=lambda x: x.get_path())
            )
        )
    )


def to_json(grouped_routes):
    """
    Convert list of :class:`RouteGroup` objects in `grouped_routes` to JSON.

    Args:
        grouped_routes (list): Llist of :class:`RouteGroup` objects.

    Returns:
        str: JSON representation of `grouped_routes`.
    """
    routes = map(
        lambda x: x.to_dict(),
        grouped_routes
    )

    return json.dumps(
        routes,
        indent=4,
        separators=(',', ': ')
    )


def gui(path="/"):
    """
    Run `bootle-gui` at given `path`.

    Args:
        path (str, default "/"): Bottle path on which the application will be
             available.

    Returns:
        fn reference: Function, which provides the `bottle-gui` functionality,\
                      mapped to bottle `path`.
    """
    @route(path)
    def root():
        """
        Handle requests to root of the project.
        """
        grouped_routes = group_routes(
            filter(lambda x: x.path not in BLACKLIST, list_routes())
        )

        accept = request.headers.get("Accept", "")
        if "json" in request.content_type.lower() or "json" in accept.lower():
            response.content_type = "application/json; charset=utf-8"
            return to_json(grouped_routes)

        return to_html(grouped_routes)

    return root


@route("/bottle_gui_static/<fn>")
def get_static(fn):
    """
    Serve static files.
    """
    return static_file(
        fn,
        root=os.path.join(os.path.dirname(__file__), 'static/')
    )


src/bottle_gui/components/napoleon2html.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
from docutils.core import publish_parts
from sphinxcontrib.napoleon import Config, GoogleDocstring


# Functions & objects =========================================================
def napoleon_to_sphinx(docstring, **config_params):
    """
    Convert napoleon docstring to plain sphinx string.

    Args:
        docstring (str): Docstring in napoleon format.
        **config_params (dict): Whatever napoleon doc configuration you want.

    Returns:
        str: Sphinx string.
    """
    if "napoleon_use_param" not in config_params:
        config_params["napoleon_use_param"] = False

    if "napoleon_use_rtype" not in config_params:
        config_params["napoleon_use_rtype"] = False

    config = Config(**config_params)

    return str(GoogleDocstring(docstring, config))


def napoleon_to_html(docstring, **config_params):
    """
    Convert `docstring` in napoleon docstring format to HTML.

    Args:
        docstring (str): Docstring in napoleon format.
        **config_params (dict): Whatever napoleon doc configuration you want.

    Returns:
        str: HTML string.
    """
    sphinxstring = napoleon_to_sphinx(docstring, **config_params)

    document = publish_parts(sphinxstring, writer_name='html')

    return document['body']


bottle_gui2.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Interpreter version: python 2.7
#
# Imports =====================================================================
import json
import os.path
from string import Template

import bottle
from bottle import route, run, static_file, request, response
import inspect

from components.napoleon2html import napoleon_to_html


# Variables ===================================================================
TEMPLATE_PATH = "static/templates/"  #: Path to the template directory.


def read_template(template_name):
    """
    Read content of the template file.

    Args:
        template_name (str): Name of the file.

    Returns:
        str: Content of `template_name` from :attr:`TEMPLATE_PATH` directory.
    """
    template_path = os.path.join(
        os.path.dirname(__file__),
        TEMPLATE_PATH,
        template_name
    )

    with open(template_path) as f:
        return f.read()


# load all necessary templates
INDEX_TEMPLATE = read_template("index.html")  #: static/templates/index.html
TABLE_TEMPLATE = read_template("table.html")  #: static/templates/table.html
ROW_TEMPLATE   = read_template("row.html")  #: static/templates/row.html
DESCR_TEMPLATE = read_template("descr.html")  #: static/templates/descr.html


# Classes =====================================================================
class RouteInfo(object):
    """
    Container for informations about `route`.

    Attributes:
        method (fn reference): Reference to undecorated function.
        path (str): Path to the function in bottle.
        args (list): Args of the function.
        docstring (str): Docstring for the function.
        mdocstring (str): Docstring for the module where the function is.
        module_name (str): Name of the module where the function is.
    """
    def __init__(self, method, path, args, docstring, mdocstring, module_name):
        """
        Attributes
            method (fn reference): see Attributes section for details.
            path (str): see Attributes section for details.
            args (list): see Attributes section for details.
            docstring (str): see Attributes section for details.
            mdocstring (str): see Attributes section for details.
            module_name (str): see Attributes section for details.
        """
        self.method = method
        self.path = path
        self.args = args
        self.docstring = self._sanitize(docstring)
        self.mdocstring = self._sanitize(mdocstring)
        self.module_name = module_name

    def _sanitize(self, s):
        """
        Replace ``<`` and ``>`` with corresponding HTML entities.

        Args:
            s (str): Input string.

        Returns:
            str: String with entities, or `s` ``if not s``.
        """
        if s:
            return s.replace("<", "&lt;").replace(">", "&gt;")

        return s

    def to_html(self):
        """
        Convert informations about this route to HTML.

        Note:
            :attr:`DESCR_TEMPLATE` and :attr:`ROW_TEMPLATE` is used.

        Returns:
            str: HTML representation of the `route`.
        """
        descr = ""

        # process docstring
        if self.docstring:
            docstring = self.docstring.strip() or ""

            descr = Template(DESCR_TEMPLATE).substitute(
                method_description=napoleon_to_html(docstring)
            )

        # wrap arguments to the html
        args = self.args or ""
        if args:
            args_style = "&#8672; &lt;<span class='param'>"
            args = args_style + "</span>, <span class='param'>".join(args)
            args += "</span>&gt;"

        return Template(ROW_TEMPLATE).substitute(
            name=self.path,
            args=args,
            http_type=self.method,
            method_description=descr
        )

    def to_dict(self):
        """
        Return dictionary representation of the class. This method is used for
        JSON output.

        Returns:
            dict: Dictionary following keys: ``method``, ``path``, ``args``, \
                  ``docstring``, ``mdocstring``, ``module_name``.
        """
        return {
            "method": self.method,
            "path": self.path,
            "args": self.args,
            "docstring": self.docstring,
            "mdocstring": self.mdocstring,
            "module_name": self.module_name,
        }

    def __str__(self):  # TODO: remove
        return self.method + " " + self.path


class RouteGroup(object):
    """
    This object is used to group :class:`RouteInfo` objects.

    Args:
        routes (list, default []): List with :class:`RouteInfo` objects.
    """
    def __init__(self, routes=[]):
        self.routes = routes

    def get_path(self):  # TODO: shortest path
        """
        Return `path` for this group.

        Returns:
            str: Path.
        """
        if len(self.routes) == 1:
            return self.routes[0].path

        # get longest path
        route_paths = map(lambda x: x.path, self.routes)
        longest_path = max(route_paths)

        if longest_path.endswith("/"):
            return longest_path

        return os.path.dirname(longest_path)

    def get_docstring(self):
        """
        Return 'module' docstring.

        Returns:
            str: Module docstring, if defined, or blank string.
        """
        if self.routes:
            return self.routes[0].mdocstring or ""

        return ""

    def to_html(self):
        """
        Convert group and all contained paths to HTML.

        Note:
            :attr:`TABLE_TEMPLATE` is used.

        Returns:
            str: HTML.
        """
        return Template(TABLE_TEMPLATE).substitute(
            name=self.get_path(),
            description=self.get_docstring(),
            rows="\n".join(
                map(lambda x: x.to_html(), self.routes)
            )
        )

    def to_dict(self):
        """
        Convert group to dict. This method is used for JSON output.

        Returns:
            dict: {path: [routes]}

        See Also:
            RouteInfo.to_dict
        """
        return {
            self.get_path(): map(lambda x: x.to_dict(), self.routes)
        }

    def __str__(self):  # TODO: remove
        return "group: " + " ".join(map(lambda x: str(x), self.routes)) + "\n"


# Functions ===================================================================
def list_routes():
    """
    Get list of :class:`RouteInfo` objects from bottle introspection.

    Returns:
        list: :class:`RouteInfo` objects.
    """
    return map(
        lambda r: RouteInfo(
            method=r.method,
            path=r.rule.split("<")[0],
            args=r.get_callback_args(),
            docstring=inspect.getdoc(r.get_undecorated_callback()) or "",
            mdocstring=inspect.getdoc(
                inspect.getmodule(r.get_undecorated_callback())
            ),
            module_name=r.get_undecorated_callback().__module__
        ),
        bottle.default_app().routes
    )


def group_routes(ungrouped_routes):
    """
    Group list of :class:`RouteInfo` objects in `ungrouped_routes` by their
    :attr:`RouteInfo.path` properties.

    Args:
        ungrouped_routes (list): List of :class:`RouteInfo` objects.

    Returns:
        list: :class:`RouteGroup` objects.
    """
    groups = []

    # go from longest routes to shorter
    routes = sorted(
        ungrouped_routes,
        key=lambda x: len(x.path),
        reverse=True
    )

    # group and remove / routes - they would break grouping algorithm later
    root_paths = filter(lambda x: x.path == "/", routes)
    if root_paths:
        groups.append(
            RouteGroup(root_paths)
        )

        # remove / routes if present
        map(lambda x: routes.pop(), root_paths)

    processed = set()
    for route in routes:
        # skip already processed routes
        if route in processed:
            continue

        # find all routes, which starts with `route`
        same_group = filter(
            lambda x: x.path.startswith(route.path),
            routes
        )

        # skip routes without groups for later processing
        if len(same_group) <= 1:  # contains always actual route, so <= 1
            continue

        groups.append(RouteGroup(same_group))
        processed.update(same_group)

    # don't forget to uprocessed routes
    for route in set(routes) - processed:
        groups.append(
            RouteGroup([route])
        )

    return groups


def to_html(grouped_routes):
    """
    Convert list of :class:`RouteGroup` objects in `group_routes` to HTML.

    Args:
        grouped_routes (list): Llist of :class:`RouteGroup` objects.

    Returns:
        str: HTML page with routes.
    """
    return Template(INDEX_TEMPLATE).substitute(
        tables="\n".join(
            map(lambda x: x.to_html(), grouped_routes)
        )
    )


def to_json(grouped_routes):
    """
    Convert list of :class:`RouteGroup` objects in `grouped_routes` to JSON.

    Args:
        grouped_routes (list): Llist of :class:`RouteGroup` objects.

    Returns:
        str: JSON representation of `grouped_routes`.
    """
    routes = map(
        lambda x: x.to_dict(),
        grouped_routes
    )

    return json.dumps(
        routes,
        indent=4,
        separators=(',', ': ')
    )


def bottle_gui(path="/"):
    """
    Run `bootle-gui` at given `path`.

    Args:
        path (str, default "/"): Bottle path on which the application will be
             available.

    Returns:
        fn reference: Function, which provides the `bottle-gui` functionality,\
                      mapped to bottle `path`.
    """
    @route(path)
    def root():
        """
        Handle requests to root of the project.
        """
        grouped_routes = group_routes(list_routes())

        accept = request.headers.get("Accept", "")
        if "json" in request.content_type.lower() or "json" in accept.lower():
            response.content_type = "application/json; charset=utf-8"
            return to_json(grouped_routes)

        return to_html(grouped_routes)

    return root


@route("/bottle_gui_static/<fn>")
def get_static(fn):
    """
    Serve static files.
    """
    return static_file(fn, root='static/')


# TODO: remove
def main():
    run(
        server="paste",
        host="127.0.0.1",
        port=8888,
        debug=True,
        reloader=True,
        # ssl_pem="static/host.pem"
    )







# Main program ================================================================
if __name__ == '__main__':
    main()
